presto-native-execution/presto_cpp/main/PeriodicServiceInventoryManager.cpp

```diff
@@ -0,0 +1,138 @@
+  eventBaseThread_.stop();
+}
+
+void PeriodicServiceInventoryManager::sendRequest() {
+  // stop() calls EventBase's destructor which executed all pending callbacks;
+  // make sure not to do anything if that's the case
+  if (stopped_) {
+    return;
+  }
+
+  try {
+    folly::SocketAddress newAddress = coordinatorDiscoverer_->getAddress();
+    // Update service address after `updateServiceTimes()` attempts.
+    if (updateServiceTimes() > 0 && (attempts_++) % updateServiceTimes() == 0) {
+      newAddress = coordinatorDiscoverer_->updateAddress();
+    }
+    if (newAddress != serviceAddress_) {
+      LOG(INFO) << "Service Inventory changed to " << newAddress.getAddressStr()
+                << ":" << newAddress.getPort();
+      std::swap(serviceAddress_, newAddress);
+      client_ = std::make_shared<http::HttpClient>(
+          eventBaseThread_.getEventBase(),
+          serviceAddress_,
+          std::chrono::milliseconds(10'000),
+          pool_,
+          clientCertAndKeyPath_,
+          ciphers_);
+    }
+  } catch (const std::exception& ex) {
+    LOG(WARNING) << "Error occurred during updating service address: "
+                 << ex.what();
+    scheduleNext();
+    return;
+  }
+
+  auto [request, body] = httpRequest();
+
+  client_->sendRequest(request, body)
+      .via(eventBaseThread_.getEventBase())
+      .thenValue([this](auto response) {
+        auto message = response->headers();
+        if (message->getStatusCode() != http::kHttpAccepted ||
+            message->getStatusCode() != http::kHttpNoContent) {
+          ++failedAttempts_;
+          LOG(WARNING) << id_ << " failed: HTTP " << message->getStatusCode()
+                       << " - " << response->dumpBodyChain();
+        } else if (response->hasError()) {
+          ++failedAttempts_;
+          LOG(ERROR) << id_ << " failed: " << response->error();
+        } else {
+          failedAttempts_ = 0;
+          LOG(INFO) << id_ << " succeeded: HTTP " << message->getStatusCode();
+        }
+      })
+      .thenError(
+          folly::tag_t<std::exception>{},
+          [this](const std::exception& e) {
+            ++failedAttempts_;
+            LOG(WARNING) << id_ << " failed: " << e.what();
+          })
+      .thenTry([this](auto /*unused*/) { scheduleNext(); });
+}
+
+uint64_t PeriodicServiceInventoryManager::getDelayMs() {
+  if (failedAttempts_ > 0 && retryFailed()) {
+    // For failure cases, execute exponential back off to ping
+    // coordinator with max back off time cap at 'frequencyMs_'.
+    auto rng = folly::ThreadLocalPRNG();
+    return folly::futures::detail::retryingJitteredExponentialBackoffDur(
+               failedAttempts_,
+               std::chrono::milliseconds(50),
+               std::chrono::milliseconds(frequencyMs_),
+               backOffjitterParam_,
+               rng)
+        .count();
+  }
+
+  // Adds some jitter for successful cases so that all workers does not ping
+  // coordinator at the same time
```

Comment:
Shall we update the comment here? It seems to me inventory manager could be used for different purpose in the future and not just for coordinator? Thanks!